pilot/pkg/proxy/envoy/v2/mesh_discovery.go

```diff
@@ -0,0 +1,213 @@
+type DiscoveryServer struct {
+	mu             sync.Mutex
+	mesh           MeshDiscovery
+	grpcServer     *grpc.Server
+	pendingStreams map[grpc.Stream]map[string]*chan bool
+}
+
+func init() {
+	envGRPCPort := os.Getenv("PILOT_GRPC_PORT")
+	if len(envGRPCPort) > 0 {
+		gRPCPort = envGRPCPort
+	}
+}
+
+// NewDiscoveryServer creates DiscoveryServer that sources data from Pilot's internal mesh data structures
+func NewDiscoveryServer(mesh MeshDiscovery) *DiscoveryServer {
+	// TODO for now use hard coded / default gRPC options. The constructor may evolve to use interfaces that guide specific options later.
+	// Example:
+	//		grpcOptions = append(grpcOptions, grpc.MaxConcurrentStreams(uint32(someconfig.MaxConcurrentStreams)))
+	var grpcOptions []grpc.ServerOption
+
+	var interceptors []grpc.UnaryServerInterceptor
+
+	// TODO: log request interceptor if debug enabled.
+
+	// setup server prometheus monitoring (as final interceptor in chain)
+	interceptors = append(interceptors, grpc_prometheus.UnaryServerInterceptor)
+	grpc_prometheus.EnableHandlingTimeHistogram()
+
+	grpcOptions = append(grpcOptions, grpc.UnaryInterceptor(grpc_middleware.ChainUnaryServer(interceptors...)))
+
+	// get the grpc server wired up
+	grpc.EnableTracing = true
+
+	grpcServer := grpc.NewServer(grpcOptions...)
+	out := &DiscoveryServer{mesh: mesh, grpcServer: grpcServer, pendingStreams: make(map[grpc.Stream]map[string]*chan bool)}
+
+	// TODO currently only supports EDS. CDS and other gRPC services need to be plugged-in here.
+	xdsapi.RegisterEndpointDiscoveryServiceServer(grpcServer, out)
+	return out
+}
+
+// Start runs DiscoveryServer's gRPC server on the port specified by the environment variable PILOT_GRPC_PORT.
+func (s *DiscoveryServer) Start() {
+	lis, err := net.Listen("tcp", ":"+gRPCPort)
+	if err != nil {
+		log.Errorf("mesh discovery server failed to listen on port %q: %v", gRPCPort, err)
+		return
+	}
+	// Serve is a blocking call.
+	go func(gRPCServer *grpc.Server, lis net.Listener) {
+		log.Infof("attempting to start mesh discovery server on port %q", gRPCPort)
+		if err := gRPCServer.Serve(lis); err != nil {
+			log.Errorf("mesh discovery server failed to serve on port %q: %v", gRPCPort, err)
+		}
+	}(s.grpcServer, lis)
+}
+
+/***************************  Mesh EDS Implementation **********************************/
+
+// StreamEndpoints implements xdsapi.EndpointDiscoveryServiceServer.StreamEndpoints().
+func (s *DiscoveryServer) StreamEndpoints(stream xdsapi.EndpointDiscoveryService_StreamEndpointsServer) error {
+	for {
+		req, err := stream.Recv()
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		// Given that Pilot holds an eventually consistent data model, Pilot ignores any acknowledgements
+		// from Envoy, whether they indicate ack success or ack failure of Pilot's previous responses.
+		if len(req.GetVersionInfo()) > 0 || len(req.GetResponseNonce()) > 0 {
```

Comment:
You will only ever get ACKs after the initial stream is opened - this stream is dedicated to a single Envoy node. If this is the logic we want, this loop should be reversed: start a goroutine to drain the response buffers, and keep the writes in this goroutine blocking.